lib/vquic/ngtcp2.c

```diff
@@ -831,20 +911,37 @@ static int cb_h3_recv_header(nghttp3_conn *conn, int64_t stream_id,
 
   if(h3name.len == sizeof(":status") - 1 &&
      !memcmp(":status", h3name.base, h3name.len)) {
+    char line[14]; /* status line is always 13 characters long */
+    size_t ncopy;
     int status = decode_status_code(h3val.base, h3val.len);
     DEBUGASSERT(status != -1);
-    msnprintf(stream->mem, stream->len, "HTTP/3 %03d \r\n", status);
+    ncopy = msnprintf(line, sizeof(line), "HTTP/3 %03d \r\n", status);
+    result = write_data(data, stream, line, ncopy);
+    if(result) {
+      return -1;
+    }
   }
   else {
     /* store as a HTTP1-style header */
-    msnprintf(stream->mem, stream->len, "%.*s: %.*s\n",
-              h3name.len, h3name.base, h3val.len, h3val.base);
+    result = write_data(data, stream, h3name.base, h3name.len);
```

Comment:
I switched from an `msnprintf` call to these four `write_data` calls. When using `msnprintf` we need to print the header line in one go, but it might need to be split if there's some space in the receive buffer but not enough for the whole line. I could have used a separate buffer to print into and then pass that buffer to a single `write_data`, but that would require a dynamically allocated buffer and I thought these separate `write_data` calls would be better.